feat: add pipeline specialized in Maven package scan

chinyeungli opened the pull request:
- Issue: https://github.com/aboutcode-org/scancode.io/issues/1763
- Create a pipeline for Maven package

chinyeungli committed:
Add "scan_maven_package" pipeline (Working-in-progress) #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Use pom_url as the datafile_path for the dependenies #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Use empty string for datafile_path in dependencies #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Removed dup code that's already present in ScanSinglePackage #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Update the matching regex for parse_maven_filename and added test #1763

- Update format

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Refactor code and add tests #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Implement "update_package_license_from_resource_if_missing" function #1763

- Update package's license if missing while the same package has license detected in RESOURCES

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Only work with "maven" package type #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Update docstring description #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Add test for the "scan_maven_package" pipeline #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Fix the Incomplete URL substring sanitization #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

tdruez commented on scanpipe/pipelines/scan_single_package.py:
```diff
@@ -126,6 +128,23 @@ def run_scan(self):
+            if not packages or not resources:
+                return
+
+        updated_packages = update_package_license_from_resource_if_missing(
```
We should use database queries instead of manipulating complex dictionaries.

tdruez commented on scanpipe/pipes/resolve.py:
```diff
@@ -521,3 +525,235 @@ def extract_headers(input_location, extract_fields):
+    return scanned_pom_packages, scanned_pom_deps
+
+
+def update_package_license_from_resource_if_missing(packages, resources):
```
This should be query-based.

tdruez reviewed (changes requested):
- Create a new maven pipe module in place of use resolve
- Opening and loading a large file to make edits multiple times in various steps is not great.
- To be discussed: Do we need a dedicated pipeline for just an extra step? Shouldn't the original `scan_single_package` detect that it's a Maven package and apply the necessary? Any reason to keep this new logic separated?

chinyeungli committed:
Add "maven.google.com" as a maven_hosts and accept .aar file #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Reorganize code structure and code enhancement #1763

- Create a new maven pipe module
- Use database queries for update_package_license_from_resource_if_missing()
- Add tests

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Update build-in-pipelines.rst #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

Aryan-SINGH-GIT committed:
exclude github folder from deploy resource

Signed-off-by: Aryan-SINGH-GIT <aryansingh12oct2005@gmail.com>

Aryan-SINGH-GIT committed:
removing unnecessary git stat

Signed-off-by: Aryan-SINGH-GIT <aryansingh12oct2005@gmail.com>

Aryan-SINGH-GIT committed:
reformat file

Signed-off-by: Aryan-SINGH-GIT <aryansingh12oct2005@gmail.com>

Aryan-SINGH-GIT committed:
adding github in deployed_resource_path_exclusions

Signed-off-by: Aryan-SINGH-GIT <aryansingh12oct2005@gmail.com>

chinyeungli committed:
Merge remote-tracking branch 'origin/main' into 1763_pipeline_for_maven

AyanSinhaMahapatra committed:
Update minecode pipelines with stable release v0.1.1 (#2013)

Signed-off-by: Ayan Sinha Mahapatra <asmahapatra@aboutcode.org>

pranavkeerti66-hue committed:
Fix typos in docstrings for pathmap.py (#1985)

Signed-off-by: pranavkeerti66-hue <pranav.keerti66@gmail.com.>

tdruez committed:
Store the whole vulnerability data from cdx to local models (#2007)

Signed-off-by: tdruez <tdruez@aboutcode.org>

chinyeungli committed:
Added steps to handle D2D in the maven pipeline #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Update "exclude_from_diff" list #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Ruff reformatted #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Update class description #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

tdruez committed:
Add project vulnerability list view (#2018)

Signed-off-by: tdruez <tdruez@aboutcode.org>

DennisClark committed:
Update index.rst with new structure

chinyeungli committed:
Error handling if pom url not found #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
feat: Improve class file mapping #2021

Signed-off-by: Chin Yeung Li <tli@nexb.com>

AyanSinhaMahapatra committed:
Fix index page issues

Signed-off-by: Ayan Sinha Mahapatra <asmahapatra@aboutcode.org>

DennisClark committed:
Update README.rst to improve structure

AyanSinhaMahapatra committed:
Update to scancode-toolkit v32.5.0 (#2000)

Signed-off-by: Ayan Sinha Mahapatra <asmahapatra@aboutcode.org>

tdruez committed:
chore: upgrade django to version 6.x (#2052)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: upgrade python in the rtd config file (#2053)

Signed-off-by: tdruez <tdruez@aboutcode.org>

AyanSinhaMahapatra committed:
Match MacOS binary symbols to Go TypeScript symbols (#2056)

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

keshav-space committed:
Bump android_inspector to v0.2.0 (#2069)

Signed-off-by: Keshav Priyadarshi <git@keshav.space>

tdruez committed:
chore: automatically closes low-quality and AI slop PRs (#2072)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade RQ, Redis, and more to their latest version (#2073)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade Django and Redis to their latest version

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: failing test by updating the expectation fixture (#2084)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade Django to latest version (#2085)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: remove the chown service in compose file (#2086)

Signed-off-by: tdruez <tdruez@aboutcode.org>

uttam282005 committed:
fix: URL-encode programming language filter values in resource list (#2079)

Signed-off-by: uttam282005 <uttam282005@gmail.com>

Monal-Reddy committed:
fix: rename "Grammar" optional step group to "Antlr" in d2d pipeline  (#2059)

Signed-off-by: Monal-Reddy <monalreddy001@gmail.com>

tdruez committed:
feat!: replace plain-text DRF token with PBKDF2-hashed API token (#2087)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: set explicit workflow permissions and pin down actions (#2090)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: bump version to v37.0.0 for release (#2091)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: add the checkout step to pypi release workflow

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: refine gh workflows for security and consistency

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: display scio and toolkit versions in place of django version (#2101)

Signed-off-by: tdruez <tdruez@aboutcode.org>

rishabh23rohil committed:
fix missing space in scan_max_file_size help text (#2097)

Signed-off-by: Rishabh Rohil <rishabhrohil024@gmail.com>

rishabh23rohil committed:
feat: add tests for chunked and get_purls utilities (#2100)

Signed-off-by: Rishabh Rohil <rishabhrohil024@gmail.com>

tdruez committed:
feat: display layers information (created_by, comment) in tree view (#2102)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: enhance the global HTML structure (#2103)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: remove the need for context processor and improve base template (#2106)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: remove the PR quality workflow

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade the scancode-action to latest version (pinned deps) (#2124)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: XSS and duplicate file detection in file upload UI (#2129)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: incomplete URL sanitization for Slack webhook detection (#2130)

Signed-off-by: tdruez <tdruez@aboutcode.org>

OmAnand857 committed:
chore: correct typo in `pipes/clamav` (#2121)

Signed-off-by: OmAnand857 <allansmith2561@gmail.com>

tdruez committed:
feat: is_safe_url to mitigate SSRF in URL requests (#2131)

Signed-off-by: tdruez <tdruez@aboutcode.org>

aayushkdev committed:
feat: add search to resource tree (#2114)

Signed-off-by: Aayush Kumar <code@aayushk.dev>

tdruez committed:
feat: enhance tree search rendering and implementation (#2133)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: use uv and uv.lock for reproducible builds in Docker image (#2134)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: add support for sqlite in the apitoken data migration (#2138)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat!: convert Dockerfile to multi-stage build (#2139)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: bump version to v37.1.0 for release (#2140)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade scancode-action to docker base v0.1 release (#2141)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: add "base" Docker build stage (#2142)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: fallback to license_declared when loading SPDX SBOM (#2144)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat!: Switch development workflow from local venv to Docker Compose (#2145)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: relax validation for non-HTTP URLs (#2147)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: load the .env in the Docker dev compose (#2149)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade Django and cryptography to latest version (#2150)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: replace the hardcoded /var/www/html by named volume (#2160)

Signed-off-by: tdruez <tdruez@aboutcode.org>

JonoYang committed:
chore: bump minecode-pipelines version to 1.0.0 (#2170)

Signed-off-by: Jono Yang <jyang@nexb.com>

JonoYang committed:
chore: update minecode-pipelines version to 1.0.1 (#2171)

Signed-off-by: Jono Yang <jyang@nexb.com>

chinyeungli committed:
Prevent duplicate CodebaseRelation entries #1763

Some projects encountered a unique constraint violation when a resource
was already mapped:
```
duplicate key value violates unique constraint "scanpipe_codebaserelation_unique_relation"
DETAIL:  Key (from_resource_id, to_resource_id, map_type)=(1512780, 1512790, jar_to_source) already exists.
```

Signed-off-by: Chin Yeung Li <tli@nexb.com>

JonoYang committed:
Install Java and jadx in Dockerfile #1812 (#2174)

Signed-off-by: Jono Yang <jyang@nexb.com>
Signed-off-by: tdruez <tdruez@aboutcode.org>
Co-authored-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade dependencies (#2173)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: bump version to v37.2.0 for release (#2175)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: add setup-buildx step to enable GHA cache support (#2176)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: add latest-full tag and OCI annotations to Docker publish workflow

Signed-off-by: tdruez <tdruez@aboutcode.org>

chinyeungli committed:
Change the text files #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Merge branch 'main' into 1763_pipeline_for_maven

Signed-off-by: Chin Yeung <tli@nexb.com>

chinyeungli committed:
Removed "def dependency_count(self):" that's removed in current main #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

tdruez committed:
chore: refine dev compose file

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: render the compliance alert value as a colored label (#2186)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: add pipeline availability check with disabled state in UI (#2188)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: replace django-taggit with a plain Project.labels JSONField (#2189)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
feat: move scanpipe settings to local settings.py with lazy resolution (#2187)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
fix: ClamAV scan crash on codebase root directory (#2190)

Signed-off-by: tdruez <tdruez@aboutcode.org>

tdruez committed:
chore: upgrade vulnerable dependencies (#2191)

Signed-off-by: tdruez <tdruez@aboutcode.org>

pombredanne commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,165 @@
+        options = ["Java", "Kotlin", "Scala", "JavaScript"]
+        d2d_config.load_ecosystem_config(pipeline=self, options=options)
+
+        self.map_about_files()
```
I am not sure you want to call all mapping steps inside a step? Should not these be their own steps?

chinyeungli replied:
I've separated the logic into different steps

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+    return scanning_errors
+
+
+def parse_maven_filename(filename):
```
Should this be in ScanCode? Do we have that already somewhere in PurlDB too?

chinyeungli replied:
removed as this is not needed.

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+                package_ns = package.get("namespace", "")
+                package_name = package.get("name", "")
+                package_version = package.get("version", "")
+                pom_url = (
```
Using this hardcoded URL means that we cannot run that on any other repo. We need to find a way to avoid that. If we use a PURL as a project input, then the Maven repo will be part of the PURL (most of the times as default), but then we do not have to further dive into that

chinyeungli replied:
Updated code to use PURL as an input.

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+        return None, None
+
+
+def get_pom_url_list(input_source, packages):
```
```suggestion
def get_pom_urls(input_source, packages):
```

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+
+def get_pom_url_list(input_source, packages):
+    """Generate Maven POM URLs from package metadata or input source."""
+    pom_url_list = []
```
May be use:
```suggestion
    pom_urls = []
```

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+def get_pom_url_list(input_source, packages):
+    """Generate Maven POM URLs from package metadata or input source."""
+    pom_url_list = []
+    if packages:
```
Not sure the if is needed at all here?

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+    """Generate Maven POM URLs from package metadata or input source."""
+    pom_url_list = []
+    if packages:
+        for package in packages:
```
What is package here? Could we get an object rather than a dict?

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+        input_source_url = input_source.get("download_url", "")
+
+        parsed_url = urlparse(input_source_url)
+        maven_hosts = {
```
See my comment above: no hardcoding of hosts should be needed

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+            "repo.maven.apache.org",
+            "maven.google.com",
+        }
+        if input_source_url and parsed_url.netloc in maven_hosts:
```
Do not we have code to do that in FetchCode, packageurl-python and/or scancode-toolkit? Also I have a hard time to understand all the parsing with indices. Do you have an example or tests of what you are trying to accomplish?

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+def construct_pom_url_from_filename(artifact_id, version):
+    """Construct a pom.xml URL from the given Maven filename."""
+    # Search Maven Central for the artifact to get its groupId
+    url = f"https://search.maven.org/solrsearch/select?q=a:{artifact_id}&wt=json"
```
I would like to avoid using search.maven.org as this is an overlay on the main Maven central repo AND this is  not guaranteed to be there AND it does not exist for other Maven repos.

pombredanne replied:
Also how could we get only an artifact and no groupid? Where do we get only a filename and no URL?

chinyeungli replied:
Removed as this is not needed.

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+        return False
+
+
+def download_pom_files(pom_url_list):
```
It is best to avoid a _list suffix IMHO
```suggestion
def download_pom_files(pom_url):
```

pombredanne commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,268 @@
+    for pom_url in pom_url_list:
+        pom_file_dict = {}
+        try:
+            downloaded_pom = fetch.fetch_http(pom_url)
```
Could we use fetchcode to fetch instead?

chinyeungli replied:
Yes, we could, but the issue is do we want to.
Following is the code if we switch to fetchcode which seems making things complicated, OR we can move the "fetch_http" to fetchcode and use it.

```
        import os
        import fetchcode
        downloaded_pom = fetchcode.fetch(pom_url)
        location = str(downloaded_pom.location)
        path = location + ".pom"
        # The fetch function from fetchcode save the file as /tmp/name
        # without an extension. We need to add the ".pom" extension so that
        # the package scan can work properly for this file.
        os.rename(location, path)
```

pombredanne reviewed (changes requested):
At last some feedback!

chinyeungli committed:
Refactor code structure for Maven pipeline #1763

- Remove unnecessary code
- Override the default "download_inputs" action
- Accept only a single Maven PURL as input
- Perform a D2D scan by default if both binary and source are available
- Update tests

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Add more comments #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Add the missing import settings #1763

* Require for `"tab_context": {"VULNERABLECODE_URL": settings.VULNERABLECODE_URL},`

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Merge branch 'main' into 1763_pipeline_for_maven

Signed-off-by: Chin Yeung <tli@nexb.com>

chinyeungli committed:
Add the missing import settings #1763
* Require for `"tab_context": {"VULNERABLECODE_URL": settings.VULNERABLECODE_URL},`

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Update tests #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Skip test_scanpipe_scan_maven_package_single_file on macos #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Better code organization #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Remove non-used code #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Restore license notice and code enhancement #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,179 @@
+
+    def maven_d2d_steps(self):
+        if self.d2d_enabled:
+            """Run D2D steps for Maven projects."""
```
Docstrings should be after the function def

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+
+
+def check_input_and_return_purl(project):
+    """Validate the input and return a PURL."""
```
```suggestion
    """Validate the input and return a maven PURL."""
```

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+    if input_purl.type != "maven":
+        error_msg = "Only maven purl is supported."
+        raise ValueError(error_msg)
+    # Should we hande all available versions if no version is provided?
```
Let' not leave todos.

tdruez commented on scanpipe/templates/scanpipe/includes/vulnerability_id.html:
```diff
@@ -27,4 +27,4 @@
   </ul>
-{% endif %}
\ No newline at end of file
+{% endif %}
```
Unwanted change

chinyeungli replied:
This was likely auto‑corrected when I ran `make check`, but in any case the missing tag has now been added and fixed.

tdruez commented on scanpipe/templates/scanpipe/tabset/tab_vulnerabilities.html:
```diff
@@ -35,4 +35,4 @@
   </tbody>
-</table>
\ No newline at end of file
+</table>
```
Unwanted change

chinyeungli replied:
This was likely auto‑corrected when I ran `make check`, but in any case the missing tag has now been added and fixed.

tdruez commented on scanpipe/views.py:
```diff
@@ -2351,6 +2352,7 @@ class DiscoveredPackageDetailsView(
             ],
             "icon_class": "fa-solid fa-bug",
             "template": "scanpipe/tabset/tab_vulnerabilities.html",
+            "tab_context": {"VULNERABLECODE_URL": settings.VULNERABLECODE_URL},
```
Unwanted change?

chinyeungli replied:
It will be removed.

tdruez commented on scanpipe/tests/test_pipelines.py:
```diff
@@ -922,6 +924,28 @@ def mock_make_git_directory(**kwargs):
         self.assertEqual(0, project1.discoveredpackages.count())
 
+    @skipIf(sys.platform == "darwin", "Not supported on macOS")
+    def test_scanpipe_scan_maven_package_single_file(self):
```
We should have a test for the d2d option as well

tdruez replied:
Still missing

chinyeungli replied:
I've added `self.assertEqual(11, project1.codebaserelations.count())` to check for the d2d relation scan.

Note that the d2d requires the source_inspector which is not supported in macos: https://github.com/aboutcode-org/scancode.io/blob/main/uv.lock#L2984

as the source_inspector will not be installed and will throw error:
https://github.com/aboutcode-org/scancode.io/actions/runs/29320243229/job/87054870086#step:7:1074

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+logger = logging.getLogger(__name__)
+
+
+def check_input_and_return_purl(project):
```
Lack of unit tests

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+    return purl
+
+
+def fetch_inputs(purl):
```
Lack of unit tests

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+
+
+def fetch_inputs(purl):
+    # Fetch the bianry and source for the given input purl and return the
```
Typo

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+        purl_src_path = [package_src.path]
+    except Exception as e:
+        logger.info("Failed to fetch source package: %s", e)
+        purl_src_path = []
```
Code duplication
Also, why are path stored as lists?

chinyeungli replied:
The path needs to be stored as a list because it is saved to `self.from_files` and `self.to_files`. Later, the pipeline calls `self.extract_inputs_to_codebase_directory()` from deploy_to_develop.py, which expects both `self.from_files` and `self.to_files` to be of list type.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+    return purl_src_path, purl_bin_path
+
+
+def fetch_and_scan_remote_pom(input_purl, scan_output_location):
```
Lack of unit tests

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+        pom_file_dict["pom_url"] = pom_url
+    except requests.RequestException:
+        # Return an empty dictionary
+        pass
```
Try block should wrap a single instruction

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+    return pom_file_dict
+
+
+def scan_pom_file(pom_file_dict):
```
No tests

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,237 @@
+                # Replace the 'datafile_path' with empty string
+                # See https://github.com/aboutcode-org/scancode.io/issues/1763#issuecomment-3525165830
+                scanned_dep["datafile_path"] = ""
+                scanned_pom_deps.append(scanned_dep)
```
Does this need to be in the context processor?

chinyeungli replied:
code updated

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,179 @@
+    copyrights. It also performs a D2D scan if both the source and binary
+    archives are available.
+
+    The output is a summary of the scan results in JSON format.
```
Let's not use the D2D wording here since it's the main documentation for the pipeline.

chinyeungli replied:
how about 

Scan a single Maven package and perform a deployment to development relation scan.

I think we still want to include the "D2D" context (perhaps in different wording)?

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,179 @@
+        scanning_errors = fetch_and_scan_remote_pom(
+            self.purl, self.scan_output_location
+        )
+        if scanning_errors:
```
Is the "if" required?

chinyeungli committed:
Apply suggestion from @tdruez

Co-authored-by: tdruez <489057+tdruez@users.noreply.github.com>
Signed-off-by: Chin Yeung <tli@nexb.com>

chinyeungli committed:
Code enhancement based on the PR review comments #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Better code structure and added tests #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,182 @@
+        # - Kotlin
+        # - Scala
+        # - JavaScript
+        options = ["Java", "Kotlin", "Scala", "JavaScript"]
```
We can see that clearly in the options list, the comment is not useful

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,182 @@
+    source archives (if they exist), and then performs scans for package
+    metadata, declared dependencies, licenses, license clarity scores, and
+    copyrights. It also performs a D2D scan if both the source and binary
+    archives are available.
```
Replace the "D2D" wording in user documentation.

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,182 @@
+        else:
+            self.extract_inputs_to_codebase_directory()
+
+    def maven_d2d_steps(self):
```
Move the method after the `d2d_*` ones for a more natural reading hierarchy.

chinyeungli replied:
what does that mean?

tdruez replied:
Cut and paste this method after d2d_process for better readability

chinyeungli replied:
I have no problem doing that, but don’t we want the methods’ order to follow the steps’ order?

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,182 @@
+                )
+
+    def update_package_license_from_resource_if_missing(self):
+        """Update PACKAGE license from the license detected in RESOURCES if missing."""
```
"PACKAGE" and "RESOURCES" are not consistent with the existing codebase conventions.

chinyeungli replied:
Do you mean I should use word "DiscoveredPackage" and "CodebaseResource" ?

tdruez replied:
"Fill in missing package licenses using licenses detected in their resources."

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+    if input_purl.type != "maven":
+        error_msg = "Only maven purl is supported."
+        raise ValueError(error_msg)
+    # Version is required
```
Comment is not useful

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+        namespace=input_purl.namespace,
+        name=input_purl.name,
+        version=input_purl.version,
+    )
```
Why recreating a `PackageURL` from `input_purl`?

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+        # Return and do nothing if data has pom.xml
+        for file in data["files"]:
+            if "pom.xml" in file["path"]:
+                return []
```
Same issue as previously reported, does the parsing need to be wrapped in the context processor?

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+    # without an extension. We need to add the ".pom" extension so that
+    # the package scan can work properly for this file.
+    os.rename(location, path)
+    """
```
This is not the right place for discussing the implementation.
Move this to the issue or the PR, not in the code.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+    try:
+        downloaded_pom = fetch.fetch_http(pom_url)
+    except requests.RequestException:
+        # Return an empty dictionary
```
`return {}` is explicit enough the comment is not adding any value

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+    pom_file_dict["pom_file_path"] = path
+    pom_file_dict["output_path"] = path + "-output.json"
+    pom_file_dict["pom_url"] = pom_url
+    return pom_file_dict
```
```suggestion
    path = str(downloaded_pom.path)
    return {
        "pom_file_path": path,
        "output_path": f"{path}-output.json",
        "pom_url": pom_url,
    }
```

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+    )
+    if scanning_errors:
+        scan_errors.append(scanning_errors)
+    return scan_errors
```
This is confusing, could you clarify?

chinyeungli replied:
code refactored and added comment.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+    scanned_dependencies = scanned_pom_data.get("dependencies", [])
+
+    for scanned_package in scanned_packages:
+        # Replace the 'datafile_path' with the pom_url
```
Avoid comments that restates the code

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+        scanned_package["datafile_paths"] = [pom_url]
+        scanned_pom_packages.append(scanned_package)
+    for scanned_dep in scanned_dependencies:
+        # Replace the 'datafile_path' with empty string
```
Avoid comments that restates the code

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+
+def update_package_license_from_resource_if_missing(project):
+    """Populate missing licenses to packages based on resource data."""
+    from license_expression import Licensing
```
Inline import?

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,229 @@
+            if detected_lics:
+                lic_exp = " AND ".join(detected_lics)
+                declared_lic_exp = str(Licensing().dedup(lic_exp))
+                package.declared_license_expression = declared_lic_exp
```
`lics` `exp` avoid those, use the real words

tdruez commented on scanpipe/tests/pipes/test_maven.py:
```diff
@@ -0,0 +1,309 @@
+    @mock.patch("scanpipe.pipes.maven.get_pom_url")
+    @mock.patch("scanpipe.pipes.maven.download_pom_file")
+    @mock.patch("scanpipe.pipes.maven.scan_pom_file")
+    @mock.patch("scanpipe.pipes.maven.update_datafile_paths")
```
Wow, what is the purpose of this test? what are we even testing here?

chinyeungli replied:
I've broken down the `fetch_and_scan_remote_pom()` and added the new `update_scan_data()`, then created a test there.

chinyeungli committed:
Clean up code and improve formatting #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Update scanpipe/pipes/maven.py

Co-authored-by: tdruez <489057+tdruez@users.noreply.github.com>
Signed-off-by: Chin Yeung <tli@nexb.com>

tdruez commented on scanpipe/pipelines/scan_maven_package.py:
```diff
@@ -0,0 +1,183 @@
+        # from deploy_to_develop.py, which expects both self.from_files and
+        # self.to_files to be of list type.
+        self.from_files = [from_file]
+        self.to_files = [to_file]
```
`fetch_inputs` can return None for either file, so `self.from_files` becomes `[None]`, which is True. `d2d_check` then enables D2D even when only one archive exists, and `collect_input_info` may set `input_path` to `None`.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,204 @@
+        return None
+
+
+def fetch_and_scan_remote_pom(input_purl, scan_output_location):
```
The return type is inconsistent: `[]`, a list of strings, or `[scanning_errors]` where `scanning_errors` is whatever `run_scan` returns (possibly `None` or `{}`). The caller does `scanning_error.items()`, which raises `AttributeError` on strings and `None`. Also `return [scanning_errors]` wraps a false success value into a non-empty list.

Please pick one error shape and return `[]` on success.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,204 @@
+    return [scanning_errors]
+
+
+def update_scan_data(data, scanned_packages, scanned_dependencies):
+    """Update packages and dependencies data"""
+    data["packages"] = data.get("packages", []) + scanned_packages
+    data["dependencies"] = scanned_dependencies
+    return data
```
Packages are appended but existing dependencies are fully replaced. If intentional (no pom.xml in codebase implies no maven dependencies), a short comment explaining why would help, otherwise dependencies should be merged like packages.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,204 @@
+    return scanned_pom_packages, scanned_pom_deps
+
+
+def update_package_license_from_resource_if_missing(project):
```
This re-scans all project resources for every package.. The package/resource relation is already in the database, so this could be done at the queryset level, filtering `codebaseresources.has_license_expression()` by `for_packages` containment per package.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,204 @@
+    """Fetch the binary and source for the given input purl"""
+    purl_str = PackageURL.to_string(purl)
+
+    purl_bin_path = fetch_path(purl_str, "binary")
+    purl_src_path = fetch_path(f"{purl_str}?classifier=sources", "source")
```
`fetch_path` is called with a PURL string but its parameter is named `url`. Rename the parameter or clarify the docstring.

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,204 @@
+    """Fetch the url and return the location of the fetched tarball"""
+    try:
+        return fetch.fetch_url(url=url).path
+    except Exception as e:
```
`except Exception` with `logger.info` silently swallows programming errors, `logger.warning` and a narrower exception type would be safer.

tdruez commented on scanpipe/tests/pipes/test_maven.py:
```diff
@@ -0,0 +1,270 @@
+    ):
+        mock_json_load.return_value = {"files": []}
+        mock_get_pom_url.return_value = "https://example.com/test.pom"
+        mock_download_pom_file.return_value = None
```
`download_pom_file` returns `{}` on failure but the mock here returns `None`. It passes thanks to the truth check, but the test doesn't exercise the real contract. Either make the function return `None` on failure or fix the mock to return `{}`.

tdruez commented on pyproject.toml:
```diff
@@ -166,6 +166,7 @@ resolve_dependencies = "scanpipe.pipelines.resolve_dependencies:ResolveDependenc
 scan_codebase = "scanpipe.pipelines.scan_codebase:ScanCodebase"
 scan_for_virus = "scanpipe.pipelines.scan_for_virus:ScanForVirus"
 scan_single_package = "scanpipe.pipelines.scan_single_package:ScanSinglePackage"
+scan_maven_package = "scanpipe.pipelines.scan_maven_package:ScanMavenPackage"
```
`scan_maven_package` should go before `scan_single_package`

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,204 @@
+logger = logging.getLogger(__name__)
+
+
+def check_input_and_return_purl(project):
```
Since the PURL is no longer rebuilt, input qualifiers are now kept. A purl submitted with `?classifier=sources` will make `fetch_inputs` produce `?classifier=sources?classifier=sources`. Either strip qualifiers here (with a docstring note) or reject purls with qualifiers explicitly.

tdruez reviewed (changes requested):
@chinyeungli a few issues still need to be addressed. Please also add or update unit tests to cover them, so we catch regressions on those.

chinyeungli committed:
Re-order the pipelines order #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Refactor Maven pipeline for error handling and performance #1763

* Better error handling
* Fix update_scan_data to append rather than overwrite existing dependencies
* Improve perfomance for update_package_license_from_resource_if_missing()

Signed-off-by: Chin Yeung Li <tli@nexb.com>

chinyeungli committed:
Fix code/doc formatting #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>

tdruez commented on scanpipe/pipes/maven.py:
```diff
@@ -0,0 +1,207 @@
+    # Return and do nothing if data has pom.xml
+    for file in data["files"]:
+        if "pom.xml" in file["path"]:
+            return []
```
```suggestion

    # Skip fetching the remote POM if a pom.xml is already present in the codebase.
    for file_entry in data["files"]:
        if PurePosixPath(file_entry["path"]).name == "pom.xml":
            return []
```

This would be a bit cleaner, mak esure to add this import too: `from pathlib import PurePosixPath`

chinyeungli replied:
updated

tdruez reviewed (changes requested):
@chinyeungli see one last suggestion before the merge/

chinyeungli committed:
Use PurePosixPath for file checking #1763

Signed-off-by: Chin Yeung Li <tli@nexb.com>
